lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py

```diff
@@ -0,0 +1,402 @@
+        examples=["001", "208", "302"],
+        pattern=r"[0-9][0-9][0-9]",
+    )
+    mnc: str = Field(
+        description="Mobile Network Code",
+        examples=["01", "001", "999"],
+        pattern=r"[0-9][0-9][0-9]?",
+    )
+    sst: int = Field(
+        description="Slice/Service Type",
+        examples=[1, 2, 3, 4],
+        ge=0,
+        le=255,
+    )
+    sd: int = Field(
+        description="Slice Differentiator",
+        examples=[1],
+        ge=0,
+        le=16777215,
+    )
+
+
+class FivegCoreGnbProviderAppData(BaseModel):
+    """Provider application data for fiveg_core_gnb."""
+    tac: int = Field(
+        description="Tracking Area Code",
+        examples=[1],
+        ge=1,
+        le=16777215,
+    )
+    plmn: list[PLMNConfig]
+
+
+class ProviderSchema(DataBagSchema):
+    """Provider schema for fiveg_core_gnb."""
+
+    app_data: FivegCoreGnbProviderAppData
+
+
+def data_matches_provider_schema(data: dict) -> bool:
+    """Return whether data matches provider schema.
+
+    Args:
+        data (dict): Data to be validated.
+
+    Returns:
+        bool: True if data matches provider schema, False otherwise.
+    """
+    try:
+        ProviderSchema(app_data=FivegCoreGnbProviderAppData(**data))
+        return True
+    except ValidationError as e:
+        logger.error("Invalid data: %s", e)
+        return False
+
+
+class GnbAvailableEvent(EventBase):
+    """Dataclass for the `fiveg_core_gnb` request event."""
+
+    def __init__(self, handle: Handle, relation_id: int):
+        """Set relation id.
+
+        Args:
+            handle (Handle): Juju framework handle.
+            relation_id : ID of the relation.
+        """
+        super().__init__(handle)
+        self.relation_id = relation_id
+
+    def snapshot(self) -> dict:
+        """Return event data.
+
+        Returns:
+            (dict): contains the relation ID.
+        """
+        return {
+            "relation_id": self.relation_id,
+        }
+
+    def restore(self, snapshot: dict) -> None:
+        """Restore event data.
+
+        Args:
+            snapshot (dict): contains the relation ID.
+        """
+        self.relation_id = snapshot["relation_id"]
+
+
+class FivegCoreGnbProviderCharmEvents(CharmEvents):
+    """Custom events for the FivegCoreGnbProvider."""
+
+    gnb_available = EventSource(GnbAvailableEvent)
+
+
+class FivegCoreGnbProvides(Object):
+    """Class to be instantiated by provider of the `fiveg_core_gnb`."""
+
+    on = FivegCoreGnbProviderCharmEvents()  # type: ignore
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Observe relation joined event.
+
+        Args:
+            charm: Juju charm
+            relation_name (str): Relation name
+        """
+        self.relation_name = relation_name
+        self.charm = charm
+        super().__init__(charm, relation_name)
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
```

Comment:
I think we should be listening on the `relation_changed` here. On `relation_joined` I don't think the gNB name will be available.